// sono/effects/index.js
'use strict';

exports.__esModule = true;
exports.reverb = exports.phaser = exports.panner = exports.flanger = exports.filter = exports.echo = exports.distortion = exports.convolver = exports.compressor = exports.analyser = undefined;

var _analyser = require('./analyser');

var _analyser2 = _interopRequireDefault(_analyser);

var _compressor = require('./compressor');

var _compressor2 = _interopRequireDefault(_compressor);

var _convolver = require('./convolver');

var _convolver2 = _interopRequireDefault(_convolver);

var _distortion = require('./distortion');

var _distortion2 = _interopRequireDefault(_distortion);

var _echo = require('./echo');

var _echo2 = _interopRequireDefault(_echo);

var _filter = require('./filter');

var _filter2 = _interopRequireDefault(_filter);

var _flanger = require('./flanger');

var _flanger2 = _interopRequireDefault(_flanger);

var _panner = require('./panner');

var _panner2 = _interopRequireDefault(_panner);

var _phaser = require('./phaser');

var _phaser2 = _interopRequireDefault(_phaser);

var _reverb = require('./reverb');

var _reverb2 = _interopRequireDefault(_reverb);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.analyser = _analyser2.default;
exports.compressor = _compressor2.default;
exports.convolver = _convolver2.default;
exports.distortion = _distortion2.default;
exports.echo = _echo2.default;
exports.filter = _filter2.default;
exports.flanger = _flanger2.default;
exports.panner = _panner2.default;
exports.phaser = _phaser2.default;
exports.reverb = _reverb2.default;
exports.default = {
    analyser: _analyser2.default,
    compressor: _compressor2.default,
    convolver: _convolver2.default,
    distortion: _distortion2.default,
    echo: _echo2.default,
    filter: _filter2.default,
    flanger: _flanger2.default,
    panner: _panner2.default,
    phaser: _phaser2.default,
    reverb: _reverb2.default
};
